core/src/main/scala/org/apache/spark/scheduler/Stage.scala

```diff
@@ -70,6 +70,26 @@ private[spark] abstract class Stage(
   /** Pointer to the latest [StageInfo] object, set by DAGScheduler. */
   var latestInfo: StageInfo = StageInfo.fromStage(this)
 
+  // The maximum number of times to retry a stage before aborting
+  val maxStageFailures = 4
+  
+  // To avoid cyclical stage failures (see SPARK-5945) we limit the number of times that a stage
+  // may be retried.
```

Comment:
I think "cyclical" is a little confusing here.  I think you could probably leave the comment out completely, or maybe something like: "Spark is resilient to executors dying, by retrying stages on FetchFailures.  However, we don't want to retry stages indefinitely, so keep track of the number of failures so we can eventually bail"